### PR TITLE
Commands, UserState + stuff

### PR DESCRIPTION
**Pull request by AydinHassan**
WIP - Some little things I wanna change

@mikeymike take a look at this when you get some time :+1:

### CODE REVIEW

**Comment by AydinHassan** [comment]
yeah yeah @codecov-io I'm going to write the tests soon, gimme a break man

File: src/Command/VerifyCommand.php [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        }
+
+        $color = new Color;
+        $color->setForceStyle(true);
+        $terminal = new UnixTerminal;
+        $width = $terminal->getWidth();
+
+        echo $color(str_repeat("─", $width))->yellow() . "\n";
+
+        $resultRenderer = new ResultsRenderer($color, (new TerminalFactory)->fromSystem(), $this->exerciseRepository);
```
Comment (AydinHassan) [inline src/Command/VerifyCommand.php]: This needs injecting

File: program.php [hunk 2]
```diff
@@ -0,0 +1,3 @@
+<?php
+
+echo count(file($argv[0]));
```
Comment (mikeymike) [inline program.php]: delete :)

File: src/Check/FunctionRequirementsCheck.php [hunk 3]
```diff
@@ -63,34 +67,25 @@ public function check(ExerciseInterface $exercise, $fileName)
...
-                        '", "',
-                        array_map(function (FuncCall $node) {
-                            return sprintf('Function: "%s" on line: "%s"', $node->name->__toString(), $node->getLine());
-                        }, $visitor->getBannedUsages())
-                    )
-                )
-            );
+            $bannedFunctions = array_map(function (FuncCall $node) {
+                return ['function' => $node->name->__toString(), 'line' => $node->getLine()];
+                //return sprintf('Function: "%s" on line: "%s"', $node->name->__toString(), $node->getLine());
```
Comment (mikeymike) [inline src/Check/FunctionRequirementsCheck.php]: delete me too :+1:

File: src/Command/HelpCommand.php [hunk 4]
```diff
@@ -0,0 +1,19 @@
+<?php
+
+namespace PhpWorkshop\PhpWorkshop\Command;
+
+/**
+ * Class HelpCommand
+ * @author Aydin Hassan <aydin@hotmail.co.uk>
+ */
+
```
Comment (mikeymike) [inline src/Command/HelpCommand.php]: LINE WTF AYDIN

File: src/Command/HelpCommand.php [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+class HelpCommand
+{
+    /**
+     * @return void
+     */
+    public function __invoke()
+    {
+        echo "HELPPPP\n";
+    }
+}
```
Comment (mikeymike) [inline src/Command/HelpCommand.php]: NO END OF LINE FILE ?!?!

File: src/Command/MenuCommand.php [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    }
+
+    /**
+     * @return void
+     */
+    public function __invoke()
+    {
+        $this->menu->display();
+    }
+}
```
Comment (mikeymike) [inline src/Command/MenuCommand.php]: again...

File: src/Command/PrintCommand.php [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+     * @param MarkdownRenderer $markdownRenderer
+     * @param Output $output
+     */
+    public function __construct(
+        ExerciseRepository $exerciseRepository,
+        UserState $userState,
+        MarkdownRenderer $markdownRenderer,
+        Output $output
+    ) {
+        $this->markdownRenderer = $markdownRenderer;
```
Comment (mikeymike) [inline src/Command/PrintCommand.php]: alignment dude

File: src/Command/PrintCommand.php [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+            return 1;
+        }
+
+        $currentExercise = $this->userState->getCurrentExercise();
+        $exercise = $this->exerciseRepository->findByName($currentExercise);
+
+        $markDown = file_get_contents($exercise->getProblem());
+        $this->output->write($this->markdownRenderer->render($markDown));
+    }
+}
```
Comment (mikeymike) [inline src/Command/PrintCommand.php]: DUDE!!! HAVE YOU EVEN GOT YOUR SNIFFER TURNED ON ????

File: src/CommandDefinition.php [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+     *
+     * @param string $name
+     * @param array $args
+     * @param string|callable $commandCallable
+     */
+    public function __construct($name, array $args, $commandCallable)
+    {
+        $this->name = $name;
+        $this->args = $args;
+        $this->commandCallable = $commandCallable;
```
Comment (mikeymike) [inline src/CommandDefinition.php]: alignment G

File: src/ExerciseRenderer.php [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+    public function __construct(
+        $appName,
+        ExerciseRepository $exerciseRepository,
+        UserState $userState,
+        UserStateSerializer $userStateSerializer,
+        MarkdownRenderer $markdownRenderer,
+        Color $color,
+        Output $output
+    ) {
+        $this->appName = $appName;
```
Comment (mikeymike) [inline src/ExerciseRenderer.php]: alignment

File: src/Result/StdOutFailure.php [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+ * @package PhpWorkshop\PhpWorkshop\Result
+ * @author  Aydin Hassan <aydin@hotmail.co.uk>
+ */
+class StdOutFailure extends Failure
+{
+    /**
+     * @var
+     */
+    private $expectedOutput;
+    /**
```
Comment (mikeymike) [inline src/Result/StdOutFailure.php]: line...

File: src/ResultRenderer/ResultsRenderer.php [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+        $lines = $this->styleArray($lines, ['green', 'bg_black']);
+
+        $lines[] = '';
+        $lines[] = $this->color->__invoke(" PASS!")->fg('green')->bg('default')->bold()->__toString();
+        $lines[] = '';
+
+        $lines[] = "Here's the official solution in case you want to compare notes:";
+        $lines[] = $this->color->__invoke(str_repeat("─", $this->terminal->getWidth()))->fg('yellow')->__toString();
+
+        $syntaxHighlighter = (new Factory)->__invoke();
```
Comment (mikeymike) [inline src/ResultRenderer/ResultsRenderer.php]: Inject this.

File: src/ResultRenderer/ResultsRenderer.php [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+        $code = explode("\n", $syntaxHighlighter->highlight(file_get_contents($exercise->getSolution())));
+        $lines = array_merge($lines, $code);
+        $lines[] = $this->color->__invoke(str_repeat("─", $this->terminal->getWidth()))->fg('yellow')->__toString();
+
+        $completedCount = count($userState->getCompletedExercises());
+        $numExercises = $this->exerciseRepository->count();
+
+
+        $lines[] = sprintf('You have %d challenges left.', $numExercises - $completedCount);
+        $lines[] = sprintf('Type "%s" to show the menu.', $_SERVER['argv'][0]);
```
Comment (mikeymike) [inline src/ResultRenderer/ResultsRenderer.php]: Server variable...

File: src/UserStateSerializer.php [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+        $data = file_get_contents($this->path);
+
+        if (trim($data) === "") {
+            $this->wipeFile();
+            return new UserState();
+        }
+
+        $json = @json_decode($data, true);
+
+        if ($json === null && json_last_error() !== JSON_ERROR_NONE) {
```
Comment (mikeymike) [inline src/UserStateSerializer.php]: yoda

File: src/Command/HelpCommand.php (hunk 5 above)
Comment (AydinHassan) [inline src/Command/HelpCommand.php]: weird, there is a new line here and most places

**Pull request merged**